pkg/controller/networkpolicy/clusternetworkpolicy.go

```diff
@@ -510,6 +510,8 @@ func (n *NetworkPolicyController) processClusterAppliedTo(appliedTo []crdv1alpha
 		var atg string
 		if at.Group != "" {
 			atg = n.processAppliedToGroupForCG(at.Group)
+		} else if at.Service != nil {
+			atg = n.createAppliedToGroupForService(at.Service)
```

Comment:
I don't think of a good reason why creating an AppliedToGroup for a service is better than transmitting the service name to agent transparently.
We create AppliedToGroups and AddressGroups to avoid transmitting whole Pods selected by the selectors every time there is an update and to enable sharing between networkpolicies. For IPBlocks, we just transmit it to agents directly. I think Service is similar to IPBlocks.
If it transmits Service to agents directly, I think you can create an index `AppliedToServicesIndex` for rule cache, so `processServiceGroupIDUpdate` can be more efficient, without having to traverse all AppliedToGroups to find affected rules.

Reply:
Thanks for the suggestion. Make sense to me.

Reply:
There are two ways to implement this: For controlplane.NetworkPolicy,

1. Add a new field parallel to the `appliedToGroups []string`

```
type NetworkPolicy struct {
	...
	AppliedToGroups []string
+       AppliedToEntities []AppliedToEntity
	...
}
type AppliedToEntities struct {
        Services []string // A list of Service namespacedNames
}

```

1. Change `appliedToGroups []string` to a new struct `AppliedToPeer` like `NetworkPolicyPeer` in rules

```
type NetworkPolicy struct {
	...
-       AppliedToGroups []string
+       AppliedTo AppliedToPeer
	...
}
type AppliedToPeer struct {
	AppliedToGroups []string
	Services []string
}

```

I'd like to say, IMO, approach 2 looks cleaner than 1, but it need to bump up the controlplane API version. Glad to hear your ideas.

Reply:
Sorry, I didn't realize we just have a single `AppliedToGroups` slice in the NetworkPolicy spec, which makes it hard to extend without breaking backwards compatability or adding a same level AppliedTo target in both Spec and Rule level. The change was beyond my expectation and I thought it over again, found we already have `PodReference`, `ExternalEntityReference`, and `NodeReference` in `GroupMember`, and thought your original approach may make more sense because:
1. It's more consistent with how we transmit object references to agents, so we wouldn't have to process 4 AppliedTo targets in a policy (2 in spec, 2 in rule), which would be more complicated.
2. Although it doesn't have performance advantage for current usage, it may be more extensible if users ask for a ServiceSelector in the future as we just need to fill selected services in the group.

Sorry for the change.

Reply:
NP. Reverted.